File: backend/app/schemas/eval.py
"""Schemas for evaluation APIs."""

from datetime import datetime
from typing import Any, Optional, Dict, List

from pydantic import BaseModel, Field, ConfigDict


class RequestTemplate(BaseModel):
    """Schema for request template configuration."""

    model_config = ConfigDict(protected_namespaces=(), extra='allow')

    model: str = "${model_name}"
    messages: List[Dict[str, Any]] = []


class EvalTaskCreate(BaseModel):
    """Schema for creating an evaluation task."""

    model_config = ConfigDict(protected_namespaces=())

    set_id: str
    model_id: str
    concurrency: int = 1  # 并发数量，默认为1（串行）
    evaluator_types: List[str] = ["exact_match"]
    request_template: Optional[RequestTemplate] = None
    system_prompt: Optional[str] = None


class EvalTaskUpdate(BaseModel):
    """Schema for updating an evaluation task."""

    model_config = ConfigDict(protected_namespaces=())

    model_id: Optional[str] = None
    concurrency: Optional[int] = None
    request_template: Optional[RequestTemplate] = None
    system_prompt: Optional[str] = None


class EvalTaskResponse(BaseModel):
    """Schema for evaluation task response."""

    model_config = ConfigDict(from_attributes=True, protected_namespaces=())

    id: str
    set_id: str
    model_id: str
    concurrency: int = 1  # 并发数量
    request_template: Optional[Dict[str, Any]] = None
    system_prompt: Optional[str] = None
    model_info: Optional[Dict[str, Any]] = None  # 模型配置详情
    status: str
    summary: Optional[Dict[str, Any]] = None
    created_at: datetime
    updated_at: datetime


class EvalResultResponse(BaseModel):
    """Schema for evaluation result response."""

    model_config = ConfigDict(from_attributes=True, protected_namespaces=())

    id: str
    run_id: str
    task_id: str
    case_id: str
    case_uid: Optional[str] = None  # 用例编号，便于展示
    actual_output: Optional[str] = None
    is_passed: bool
    execution_error: Optional[str] = None  # 执行错误信息，如模型调用失败
    evaluator_logs: List[Any]
    created_at: datetime


class EvalRunResponse(BaseModel):
    """Schema for evaluation run response."""

    model_config = ConfigDict(from_attributes=True, protected_namespaces=())

    id: str
    task_id: str
    run_number: int
    status: str
    summary: Optional[Dict[str, Any]] = None
    started_at: datetime
    completed_at: Optional[datetime] = None
    error: Optional[str] = None


class EvalProgressEvent(BaseModel):
    """Schema for SSE progress event."""

    type: str
    data: Dict[str, Any]


class EvalSummary(BaseModel):
    """Schema for evaluation summary."""

    total: int = 0
    passed: int = 0
    failed: int = 0
    pass_rate: float = 0.0


class TemplateTestRequest(BaseModel):
    """Schema for testing request template."""

    case_id: Optional[str] = None
    test_input: Optional[str] = None


class TemplateTestResponse(BaseModel):
    """Schema for template test response."""

    rendered_request: Dict[str, Any]
    actual_response: Optional[str] = None
    error: Optional[str] = None

File: frontend/src/types/eval.ts
/** Type definitions for evaluation */

export interface ModelInfo {
  model_code: string
  display_name: string
  provider: {
    id: string
    name: string
  }
}

export interface RequestTemplate {
  model: string
  messages: Array<{ role: string; content: string }>
  [key: string]: any
}

export interface EvalTaskCreate {
  name?: string
  set_id: string
  model_id: string
  concurrency?: number  // 并发数量，默认1
  evaluator_types?: string[]
  request_template?: RequestTemplate
  system_prompt?: string
}

export interface EvalTaskUpdate {
  name?: string
  model_id?: string
  concurrency?: number  // 并发数量
  request_template?: RequestTemplate
  system_prompt?: string
}

export interface TemplateTestRequest {
  case_id?: string
  test_input?: string
}

export interface TemplateTestResponse {
  rendered_request: Record<string, any>
  actual_response?: string
  error?: string
}

export interface EvalTask {
  id: string
  name?: string | null
  set_id: string
  model_id: string
  concurrency: number  // 并发数量
  request_template?: RequestTemplate
  system_prompt?: string
  model_info?: ModelInfo
  status: 'PENDING' | 'RUNNING' | 'COMPLETED' | 'FAILED'
  summary: EvalSummary | null
  created_at: string
  updated_at: string
}

export interface EvalSummary {
  total: number
  passed: number
  failed: number
  pass_rate: number
}

export interface EvalResult {
  id: string
  run_id: string
  task_id: string
  case_id: string
  case_uid?: string | null  // 用例编号，便于展示
  actual_output: string | null
  is_passed: boolean
  execution_error?: string | null  // 执行错误信息
  evaluator_logs: EvaluatorLog[]
  execution_duration?: number | null  // 执行时长（毫秒）
  skill_tokens?: number | null  // 技能LLM调用的token消耗
  evaluator_tokens?: number | null  // LLM评估器的token消耗
  created_at: string
}

export interface EvalRun {
  id: string
  task_id: string
  run_number: number
  status: 'PENDING' | 'RUNNING' | 'COMPLETED' | 'FAILED'
  summary: EvalSummary | null
  started_at: string
  completed_at: string | null
  error: string | null
  total_duration_ms?: number | null  // 总执行时长（毫秒）
  total_skill_tokens?: number | null  // 总技能LLM tokens消耗
  total_evaluator_tokens?: number | null  // 总评估器LLM tokens消耗
}

export interface EvaluatorLog {
  evaluator: string
  passed: boolean
  reason?: string
}

export interface EvalProgressEvent {
  type: 'progress' | 'result' | 'complete' | 'error'
  data: Record<string, any>
}
